lib/providers/ipmeta_provider_maxmind_v2.c

```diff
@@ -0,0 +1,914 @@
+      state->parser.status = CSV_EUSER;
+      return;
+    }
+    }
+    break;
+
+  case BLOCKS_COL_ACCURACY:
+    /* accuracy */
+    if (tok != NULL) {
+      tmp->accuracy = strtol(tok, &end, 10);
+      if (*tok != '\0' && (end == tok || *end != '\0' || errno == ERANGE)) {
+        ipmeta_log(__func__, "Invalid Area Code Value (%s)", tok);
+        state->parser.status = CSV_EUSER;
+        return;
+      }
+    }
+    break;
+
+  default:
+    ipmeta_log(__func__, "Invalid Blocks Column (%d:%d)", state->current_line,
+               state->current_column);
+    state->parser.status = CSV_EUSER;
+    break;
+  }
+
+  /* move on to the next column */
+  state->current_column++;
+}
+
+static void parse_blocks_row(int c, void *data)
+{
+  ipmeta_provider_t *provider = (ipmeta_provider_t *)data;
+  ipmeta_provider_maxmind_v2_state_t *state = STATE(provider);
+  /* Initializing block record*/
+  ipmeta_record_t *block_record = &(state->tmp_record);
+  /* the pointer to the final record */
+  ipmeta_record_t *record = NULL;
+  /*location record */
+  ipmeta_record_t *loc_record = NULL;
+  khiter_t khiter;
+
+
+  if (state->current_line < HEADER_ROW_CNT) {
+    state->current_line++;
+    return;
+  }
+
+  /* done processing the line */
+
+  /* make sure we parsed exactly as many columns as we anticipated */
+  if (state->current_column != BLOCKS_COL_COUNT) {
+    ipmeta_log(__func__,
+               "ERROR: Expecting %d columns in the blocks file, "
+               "but actually got %d",
+               BLOCKS_COL_COUNT, state->current_column);
+    state->parser.status = CSV_EUSER;
+    return;
+  }
+
+/* At the end of each row, look up the locations row in the hash table */
+
+// We can only make one memcopy of the locations block into the final block 
+  if ((record = ipmeta_provider_init_record(provider, state->current_line)) ==
+      NULL) {
+    ipmeta_log(__func__, "ERROR: Could not initialize meta record");
+    state->parser.status = CSV_EUSER;
+    return;
+  }
+
+if (block_record->id != 0){
+  if ((khiter = kh_get(loctemp_rcd, state->locations, block_record->id)) ==
+      kh_end(state->locations)) {
+    ipmeta_log(__func__, "ERROR: Invalid geoname id (%"PRIu32")",
+                block_record->id);
+    state->parser.status = CSV_EUSER;
+    return;
+  }
+
+  // copy anything in the locations row into the temp record.
+  loc_record = &(kh_value(state->locations, khiter));
+  memcpy(record,loc_record, sizeof(ipmeta_record_t));
+
+  // Trying to duplicate the char values we avoid a free memory error.  
+  if (loc_record->region != NULL){
+  if((record->region = strdup(loc_record->region)) == NULL) {
+    ipmeta_log(__func__, "ERROR: Failed to duplicate region");
+  }
+  }
+
+  /* It is essential to duplicate blocks so that we do not end up with a free 
+  non-allocated memory case. */
+  if (loc_record->city != NULL){
+    if((record->city= strdup(loc_record->city)) == NULL) {
```

Comment:
these can be simplified into sth like `if (X != NULL && (X = strdup(Y)) == NULL)`